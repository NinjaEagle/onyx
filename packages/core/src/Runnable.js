/**
 * `Runnable` class that abstracts basic functionality shared
 * between the `Test` and `Suite` classes.
 * 
 * @class
 * @param {String} title
 * @param {Function} fn
 */
export default class Runnable {
  title = '';
  fn = null;
  pending = null;
  sync = null;
  async = !sync;

  constructor(title, fn) {
    this.title = title;
    this.fn = fn;
    this.async = fn && fn.length;
    this.sync = !this.async;
    this.pending = false;
  }


  /**
   * Check if this runnable or its parent suite is marked as pending.
   *
   * @private
   */
  _isPending() {
    return this.pending || this.isPending();
  }
}
