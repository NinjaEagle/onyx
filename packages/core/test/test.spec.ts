import Runnable from '../src/runnable';
import Suite from '../src/suite';
import Test, { isTest } from '../src/test';

describe('Test', () => {
  it('should run', () => {
    const fn = jest.fn();
    const test = new Test('test', fn);

    expect(fn).toHaveBeenCalledTimes(0);
    expect(test.run()).toMatchSnapshot();
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should run asynchronously', async () => {
    const fn = jest.fn();
    const test = new Test('test', fn);

    expect(fn).toHaveBeenCalledTimes(0);
    expect(await test.asyncRun()).toMatchSnapshot();
    expect(fn).toHaveBeenCalledTimes(1);
  });

  it('should return isDone', () => {
    const test = new Test('test', () => null);

    expect(test.isDone()).toBeFalsy();
    test.run();
    expect(test.isDone()).toBeTruthy();
  });

  it('should skip', () => {
    const fn = jest.fn();
    const test = new Test('test', fn, true);

    expect(test.run()).toMatchSnapshot();
    expect(fn).toHaveBeenCalledTimes(0);
  });

  it('should fail', () => {
    const fn = () => {
      throw new Error('Fatal error');
    };
    const test = new Test('test', fn);

    expect(test.run()).toMatchSnapshot();
  });

  it('should fail in async', async () => {
    const err = new Error('Fatal error');
    const fn = () => {
      throw err;
    };
    const test = new Test('test', fn);

    const spy = jest.fn();
    await test.asyncRun().catch(spy);

    expect(spy).toHaveBeenCalledWith(err);
  });

  it('should check if is test', () => {
    expect(isTest(null)).toBeFalsy();
    expect(isTest({})).toBeFalsy();
    expect(isTest(new Runnable('not a test'))).toBeFalsy();
    expect(isTest(new Suite('not a test'))).toBeFalsy();
    expect(isTest(new Test('a test', () => null))).toBeTruthy();
  });
});
