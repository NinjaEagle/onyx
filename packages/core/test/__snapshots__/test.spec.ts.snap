// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test async should run 1`] = `
Result {
  "internalMessages": Array [],
  "internalStatus": "passed",
}
`;

exports[`Test should fail 1`] = `
Result {
  "internalMessages": Array [
    "Error: Fatal error",
  ],
  "internalStatus": "failed",
}
`;

exports[`Test should run 1`] = `
Result {
  "internalMessages": Array [],
  "internalStatus": "passed",
}
`;

exports[`Test should skip 1`] = `
Result {
  "internalMessages": Array [],
  "internalStatus": "skipped",
}
`;

exports[`Test should skip 2`] = `
Result {
  "internalMessages": Array [],
  "internalStatus": "todo",
}
`;
